coding_standard_c.txt

```diff
@@ -0,0 +1,147 @@
+
+Files
+Typically, every header file has the same base filename as an associated source file.  Each header file is bound by conditional compilation commands using an all caps and underscored variant of the complete filename.  Header files should never define functions or variables.
+
+/*
+ * File: basic_file.h
+ * Header file block comment
+ */
+#ifndef BASIC_FILE_H
+#define BASIC_FILE_H
+/* standard #includes */
+/* project #includes */
+/* type/constant definitions */
+/* extern variable declarations */
+/* extern function declarations */
+#endif /* BASIC_FILE_H */
+/*
+ * File: basic_file.c
+ * Source file block comment...
+ */
+/* standard #includes */
+/* project #includes */
+#include “basic_file.h”
+/* externalized variable definitions */
+/* static variables (locally global) */
+/* static function declarations */
+/* externalized function definitions */
+/* static function definitions */
+
+Header files should only #include what is necessary to allow a file that includes it to compile.  Associated source files will always #include the header of the same name, but should #include files whose resources are used within the source even if they are already included in that header.  This provides a complete context for readers of the source file... i.e., they don't have to search through headers to determine  where a resource came from.  For example, consider a header and source that use FILE pointers declared in stdio.h.  The header must #include <stdio.h> to ensure files that include it will be able to compile.  The corresponding source #includes the associated header and therefore receives the benefit of stdio.h automatically.  The source would be unnerving if the FILE type was used and stdio.h was not #included.
+
+Variable Scope and types
+Variables should be declared and scoped to the smallest scope of use.
+
+/* This should be avoided */
+void foo() {
+	unsigned i;
+	do_something();
+	for(i=0; i < 100; i++) {
+		do_more(i);
+	}
+}
+
+/* Proper variable scope */
+void foo() {
+
+	do_something();
+	unsigned i;
+	for(i=0; i < 100; i++) {
+		do_more(i);
+	}
+}
+
+Variables should only be of a signed type if something should ever be negative. A common, incorrect use, is to declare loop indices as int instead of unsigned.
+
+Formatting
+Always use tab characters rather than spaces at the beginning of lines.  Make sure your editing environment uses at least five visual spaces to represent tabs.  This will make code push horizontal limits with very little depth, thus deterring extraordinary amounts of nested statements.  The large tab approach “makes poorly-written programs blatantly ugly”. [2]  It may periodically cause small pieces of well written code to seem unruly, but it is worth it overall.
```

Comment:
Tab indentation instead of spaces is going to be hard to sell here I think. I'm pretty attached to indenting with an explicit number of spaces (4 typically). This will also cause a pile of space to tab whitespace churn that will be really weird. Are you flexible on this one?